File: src/main/java/org/cubexmc/metro/gui/ChatInputManager.java
package org.cubexmc.metro.gui;

import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.util.SchedulerUtil;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class ChatInputManager implements Listener {

    private final Metro plugin;
    private final ChatCallbackScheduler callbackScheduler;
    private final Map<UUID, ChatInputCallback> pendingInputs = new HashMap<>();

    public ChatInputManager(Metro plugin) {
        this(plugin, (metro, player, task) -> SchedulerUtil.entityRun(metro, player, task, 0L, -1L));
    }

    ChatInputManager(Metro plugin, ChatCallbackScheduler callbackScheduler) {
        this.plugin = plugin;
        this.callbackScheduler = callbackScheduler;
    }

    public void requestInput(Player player, String prompt, ChatInputCallback callback) {
        player.closeInventory();
        player.sendMessage(prompt);
        pendingInputs.put(player.getUniqueId(), callback);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerChat(AsyncPlayerChatEvent event) {
        Player player = event.getPlayer();
        UUID uuid = player.getUniqueId();
        
        if (pendingInputs.containsKey(uuid)) {
            event.setCancelled(true);
            String input = event.getMessage();
            ChatInputCallback callback = pendingInputs.remove(uuid);
            
            if (input.equalsIgnoreCase("cancel") || input.equalsIgnoreCase("取消")) {
                player.sendMessage(plugin.getLanguageManager().getMessage("chat.input_cancelled"));
                scheduleCallback(player, callback::onCancel);
                return;
            }
            
            scheduleCallback(player, () -> callback.onInput(input));
        }
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        pendingInputs.remove(event.getPlayer().getUniqueId());
    }

    public interface ChatInputCallback {
        void onInput(String input);
        default void onCancel() {}
    }

    @FunctionalInterface
    interface ChatCallbackScheduler {
        void run(Metro plugin, Player player, Runnable task);
    }

    private void scheduleCallback(Player player, Runnable task) {
        callbackScheduler.run(plugin, player, task);
    }
}

File: src/main/java/org/cubexmc/metro/gui/ChatInputManager.kt
package org.cubexmc.metro.gui

import java.util.UUID
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.cubexmc.metro.Metro
import org.cubexmc.metro.util.SchedulerUtil

class ChatInputManager(
    private val plugin: Metro,
    private val callbackScheduler: ChatCallbackScheduler,
) : Listener {
    constructor(plugin: Metro) : this(
        plugin,
        ChatCallbackScheduler { metro, player, task -> SchedulerUtil.entityRun(metro, player, task, 0L, -1L) },
    )

    private val pendingInputs: MutableMap<UUID, ChatInputCallback> = HashMap()

    fun requestInput(player: Player, prompt: String, callback: ChatInputCallback) {
        player.closeInventory()
        player.sendMessage(prompt)
        pendingInputs[player.uniqueId] = callback
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerChat(event: AsyncPlayerChatEvent) {
        val player = event.player
        val uuid = player.uniqueId
        if (!pendingInputs.containsKey(uuid)) return

        event.isCancelled = true
        val input = event.message
        val callback = pendingInputs.remove(uuid) ?: return
        if (input.equals("cancel", ignoreCase = true) || input.equals("取消", ignoreCase = true)) {
            player.sendMessage(plugin.languageManager.getMessage("chat.input_cancelled"))
            scheduleCallback(player) { callback.onCancel() }
            return
        }
        scheduleCallback(player) { callback.onInput(input) }
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        pendingInputs.remove(event.player.uniqueId)
    }

    interface ChatInputCallback {
        fun onInput(input: String)
        fun onCancel() {}
    }

    fun interface ChatCallbackScheduler {
        fun run(plugin: Metro, player: Player, task: Runnable)
    }

    private fun scheduleCallback(player: Player, task: Runnable) {
        callbackScheduler.run(plugin, player, task)
    }
}
